src/main/java/net/minecraftforge/common/DimensionManager.java

```diff
@@ -286,16 +316,39 @@ public static WorldProvider createProviderFor(int dim)
         }
     }
 
-    public static void unloadWorld(int id) {
-        unloadQueue.add(id);
+    public static void unloadWorld(int id)
+    {
+        if(!unloadQueue.contains(id))
+        {
+            FMLLog.fine("Queueing dimension %s to unload", id);
+            unloadQueue.add(id);
+        }
+        else
+        {
+            dimensions.get(id).ticksWaited = 0;
+        }
     }
 
     /*
     * To be called by the server at the appropriate time, do not call from mod code.
     */
     public static void unloadWorlds(Hashtable<Integer, long[]> worldTickTimes) {
+        ArrayList<Integer> toRemove = new ArrayList<Integer>();
         for (int id : unloadQueue) {
+            Dimension dimension = dimensions.get(id);
+            if (dimension.ticksWaited != dimension.delayToUnload)
+            {
+                dimension.ticksWaited++;
+                continue;
+            }
             WorldServer w = worlds.get(id);
+            toRemove.add(id);
+            dimension.ticksWaited = 0;
+            if (w != null && (!ForgeChunkManager.getPersistentChunksFor(w).isEmpty() || !w.playerEntities.isEmpty()) || dimension.type.shouldLoadSpawn()) //Don't unload the world if the status changed
+            {
+                FMLLog.fine("Aborting unload for dimension %s as status changed", id);
+                continue;
+            }
             try {
                 if (w != null)
```

Comment:
Its not multithreaded AFAIK, but the question is if we should support modders unloading worlds in their own code